### PR TITLE
Create updateContribution Handler

### PR DESCRIPTION
**Pull request by topocount**
## Summary

The update contribution handler ensures a contribution can't be changed
to a time that was included in a closed epoch. Contributions with a
timestamp in closed epochs are not editable. Contributions cannot have
empty descriptions. Contributions must exist for the user as well.

## Permissions Updates

Make inserts allowable from the autogenerated api. It's impossible to spam
contributions thanks to the custom check.

completely disallow updates from the autogenerated api since the action
exists.

allow selects for all contributions added by the user.

## Test Plan

Added unit tests to ensure all error cases are handled

integration tests will be written when the frontend is implemented (imminent)

## Merge Plan

Merge after #1327 and rebase on top of it.

### CODE REVIEW

File: api/hasura/actions/_handlers/updateContribution.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  const { id, description, datetime_created } = payload;
+
+  const contribution = await fetchAndVerifyContribution({
+    id,
+    res,
+    userAddress,
+    operationName: actionName,
+  });
+
+  if (!contribution) return;
```
Comment (zashton) [inline api/hasura/actions/_handlers/updateContribution.ts:33]: seems to be missing an error response

File: hasura/metadata/actions.graphql [hunk 2]
```diff
@@ -1,155 +1,111 @@
...
-    payload: UpdateCircleInput!
-  ): UpdateCircleOutput
+  updateCircle(payload: UpdateCircleInput!): UpdateCircleOutput
 }
 
 type Mutation {
-  updateEpoch(
-    payload: UpdateEpochInput!
-  ): EpochResponse
+  updateContribution(
```
Comment (zashton) [inline hasura/metadata/actions.graphql:74]: the formatting seem to be off here

File: hasura/metadata/actions.graphql [hunk 3]
```diff
@@ -351,6 +307,12 @@ input CreateUserWithTokenInput {
   token: String!
 }
 
+input UpdateContributionInput {
```
Comment (zashton) [inline hasura/metadata/actions.graphql:310]: formatting for this one as well

**Review by zashton — CHANGES_REQUESTED** [review]

File: api/hasura/actions/_handlers/updateContribution.ts (hunk 1 above)
Comment (topocount) [reply api/hasura/actions/_handlers/updateContribution.ts:33]: @zashton the error responses are inside the above function. If no contribution is returned, we cannot continue and must return.

File: api/hasura/actions/_handlers/updateContribution.test.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    await handler(req, res);
+
+    expect(res.status).toHaveBeenCalledWith(422);
+    expect(res.json).toHaveBeenCalled();
+    const results = (res.json as any).mock.calls[0][0];
+    expect(results.message).toBe(
+      'cannot reassign contribution to a closed epoch'
+    );
+  });
+  test('cannot delete a contribution from a different user', async () => {
```
Comment (zashton) [inline api/hasura/actions/_handlers/updateContribution.test.ts:131]: small nit 
```suggestion
  test('cannot update a contribution from a different user', async () => {
```

**Review by zashton — APPROVED** [review]

**Pull request merged**